src/DAG/dag_parser.c

```diff
@@ -0,0 +1,459 @@
+                currentOp->commandType = REDISAI_DAG_CMD_MODELRUN;
+                RAI_Model *mto;
+                RedisModuleKey *modelKey;
+                const int status = RAI_GetModelFromKeyspace(ctx, argv[arg_pos + 1], &modelKey, &mto,
+                                                            REDISMODULE_READ);
+                if (status == REDISMODULE_ERR) {
+                    RAI_FreeRunInfo(rinfo);
+                    RedisModule_ReplyWithError(ctx, "ERR Model not found");
+                    return REDISMODULE_ERR;
+                }
+                currentOp->devicestr = mto->devicestr;
+                currentOp->runkey = argv[arg_pos + 1];
+                currentOp->mctx = RAI_ModelRunCtxCreate(mto);
+            }
+            if (!strcasecmp(arg_string, "AI.SCRIPTRUN")) {
+                if (argc - 3 < arg_pos) {
+                    RedisModule_WrongArity(ctx);
+                    return REDISMODULE_ERR;
+                }
+                RAI_DagOp *currentOp = rinfo->dagOps[rinfo->dagOpCount];
+                currentOp->commandType = REDISAI_DAG_CMD_SCRIPTRUN;
+                RAI_Script *sto;
+                RedisModuleKey *scriptKey;
+                const int status = RAI_GetScriptFromKeyspace(ctx, argv[arg_pos + 1], &scriptKey,
+                                                             &sto, REDISMODULE_READ);
+                if (status == REDISMODULE_ERR) {
+                    RAI_FreeRunInfo(rinfo);
+                    return REDISMODULE_ERR;
+                }
+                currentOp->devicestr = sto->devicestr;
+                const char *functionName = RedisModule_StringPtrLen(argv[arg_pos + 2], NULL);
+                currentOp->runkey = argv[arg_pos + 1];
+                currentOp->sctx = RAI_ScriptRunCtxCreate(sto, functionName);
+            }
+            if (RMAPI_FUNC_SUPPORTED(RedisModule_HoldString)) {
+                RedisModule_HoldString(NULL, argv[arg_pos]);
+            } else {
+                RedisModule_RetainString(NULL, argv[arg_pos]);
+            }
+            RAI_DagOp *currentOp = rinfo->dagOps[rinfo->dagOpCount];
+            currentOp->argv = array_append(currentOp->argv, argv[arg_pos]);
+            currentOp->argc++;
+        }
+    }
+
+    rinfo->dagOpCount = array_len(rinfo->dagOps);
+
+    for (long long i = 0; i < array_len(rinfo->dagOps); i++) {
+        RAI_DagOp *currentOp = rinfo->dagOps[i];
+        if (currentOp == NULL)
+            continue;
+        int parse_result;
+        switch (currentOp->commandType) {
+        case REDISAI_DAG_CMD_TENSORSET:
+            currentOp->outkeys = array_append(currentOp->outkeys, currentOp->argv[1]);
+            break;
+        case REDISAI_DAG_CMD_TENSORGET:
+            currentOp->inkeys = array_append(currentOp->inkeys, currentOp->argv[1]);
+            break;
+        case REDISAI_DAG_CMD_MODELRUN:
+            parse_result = RedisAI_Parse_ModelRun_RedisCommand(
```

Comment:
this is the logic duplication we discussed about, right?

Reply:
Right